src/wp-includes/class-wp-block-metadata-registry.php

```diff
@@ -0,0 +1,203 @@
+<?php
+/**
+ * Block Metadata Registry
+ *
+ * @package WordPress
+ * @subpackage Blocks
+ * @since 6.X.0
+ */
+
+/**
+ * Class used for managing block metadata collections.
+ *
+ * The WP_Block_Metadata_Registry allows plugins to register metadata for large
+ * collections of blocks (e.g., 50-100+) using a single PHP file. This approach
+ * reduces the need to read and decode multiple `block.json` files, enhancing
+ * performance through opcode caching.
+ *
+ * @since 6.X.0
+ */
+class WP_Block_Metadata_Registry {
+
+	/**
+	 * Container for storing block metadata collections.
+	 *
+	 * Each entry maps a base path to its corresponding metadata and callback.
+	 *
+	 * @since 6.X.0
+	 * @var array<string, array<string, mixed>>
+	 */
+	private static $collections = array();
+
+	/**
+	 * Caches the last matched collection path for performance optimization.
+	 *
+	 * @since 6.X.0
+	 * @var string|null
+	 */
+	private static $last_matched_collection = null;
+
+	/**
+	 * Registers a block metadata collection.
+	 *
+	 * This method allows registering a collection of block metadata from a single
+	 * manifest file, improving performance for large sets of blocks.
+	 *
+	 * @since 6.X.0
+	 *
+	 * @param string   $path                The absolute base path for the collection ( e.g., WP_PLUGIN_DIR . '/my-plugin/blocks/' ).
+	 * @param string   $manifest            The absolute path to the manifest file containing the metadata collection.
+	 * @param callable $identifier_callback Optional. Callback to determine the block identifier from a path.
+	 *                                      The callback should accept a string (file or folder path) and return a string (block identifier).
+	 *                                      This allows custom mapping between file paths and block names in the manifest.
+	 *                                      If null, the default identifier callback is used, which extracts the parent
+	 *                                      directory name. For example, when calling get_metadata() with a path like
+	 *                                      'WP_PLUGIN_DIR/my-plugin/blocks/example/block.json', it would look for
+	 *                                      a key named "example" in the manifest.
+	 * @return bool                         True if the collection was registered successfully, false otherwise.
```

Comment:
While I'm the one that originally proposed this, we may want to leave this out of the initial iteration. It looks like we don't know whether anyone would need this. We can always add it later if we find such a need, but we cannot remove it once it's released. So for hypothetical things like this it's better to wait until a real need arises.

Reply:
Sounds good. I have removed it; tell me if I went too excessive in the comments for `register_collection()`. I was mostly afraid of someone new to this not understanding what's going on.